### PR TITLE
[Build] Making grpcio_tools ready for upcoming Protobuf change.

### PR DESCRIPTION
**Pull request by veblush**
Protobuf is beginning to rely more on upb, pulling in more C files. Hence, grpcio_tool needs the following changes to absorb this.

- Changing the setup.py to support both C and C++ by removing explicit language=c++. Rather it can choose the right compiler by its extension.
- Adding build_extensions injection to deal with C/C++ option conflict as grpcio does.
- Adding protobuf and upb directories to the build script so that it can find newly added source files in protobuf.

Tested by https://github.com/grpc/grpc/pull/35796

### CODE REVIEW

**Comment by gnossen** [comment]
@XuanWang-Amos Can you please review this?

File: tools/distrib/python/grpcio_tools/setup.py [hunk 1]
```diff
@@ -126,6 +126,40 @@ def get_ext_filename(self, ext_name):
...
+        #
+        #   If we are not using a permissive compiler that's OK with being
+        #   passed wrong std flags, swap out compile function by adding a filter
+        #   for it.
+        old_compile = self.compiler._compile
+
+        def new_compile(obj, src, ext, cc_args, extra_postargs, pp_opts):
+            if src.endswith(".c"):
+                extra_postargs = [
+                    arg for arg in extra_postargs if not "-std=c++" in arg
```
Comment (XuanWang-Amos) [inline tools/distrib/python/grpcio_tools/setup.py:144]: [Minor] I think it's more easy to understand by saying `if "-std=c++" not in arg`

File: tools/distrib/python/make_grpcio_tools.py [hunk 2]
```diff
@@ -71,7 +71,9 @@
 # will be added to include path when building grpcio_tools
 CC_INCLUDES = [
     os.path.join("third_party", "abseil-cpp"),
+    os.path.join("third_party", "protobuf"),
     os.path.join("third_party", "protobuf", "src"),
+    os.path.join("third_party", "protobuf", "upb"),
```
Comment (XuanWang-Amos) [inline tools/distrib/python/make_grpcio_tools.py:76]: Can we remove those two lines if we already have `os.path.join("third_party", "protobuf")`?

**Review by XuanWang-Amos — APPROVED** [review]

File: tools/distrib/python/grpcio_tools/setup.py (hunk 1 above)
Comment (veblush) [reply tools/distrib/python/grpcio_tools/setup.py:144]: Good point. I'll take care of this in a separate PR to clean this up on other usage as well.

File: tools/distrib/python/make_grpcio_tools.py (hunk 2 above)
Comment (veblush) [reply tools/distrib/python/make_grpcio_tools.py:76]: If protobuf has other directories other than src and upb as an example, having a single protobuf is copying those unnecessary files?

**Pull request closed without merging**